client/ping.go

```diff
@@ -16,15 +16,20 @@ func (cli *Client) Ping(ctx context.Context) (types.Ping, error) {
 	if err != nil {
 		return ping, err
 	}
-	defer ensureReaderClosed(serverResp)
 
-	ping.APIVersion = serverResp.header.Get("API-Version")
+	if serverResp.header != nil {
+		ping.APIVersion = serverResp.header.Get("API-Version")
 
-	if serverResp.header.Get("Docker-Experimental") == "true" {
-		ping.Experimental = true
+		if serverResp.header.Get("Docker-Experimental") == "true" {
+			ping.Experimental = true
+		}
+		ping.OSType = serverResp.header.Get("OSType")
 	}
 
-	ping.OSType = serverResp.header.Get("OSType")
+	if err := cli.checkResponseErr(serverResp); err != nil {
```

Comment:
nit: this if isn't needed, and could be just;
(also not sure if the `defer` was needed?)

```go
err := cli.checkResponseErr(serverResp)
ensureReaderClosed(serverResp)
return ping, err
```